Console Prototype

LunexCoding opened the pull request:
The following changes will focus on the early release of the console prototype.

LunexCoding committed:
Add base modules

Added base modules:
  * `ingredient`
  * `recipe`
  * `book`
  * `fredge`

Created `RecipeBook` package consisting of modules:
  * `ingredient`
  * `recipe`
  * `book`

In the plans:
  * Designing the basic architecture of the application
  * Define data format for ingredients and recipes
  * Writing unit tests
  * Synchronization with FTP server

These actions will be performed in random order, as needed.

LunexCoding committed:
Add logging and storages

Added storages:

* `storage.py` - to store information about products in stock
* `ingredientsDB.py` - base for operations with products in memory

Added `fileSystem.py`.

LunexCoding committed:
Saving data to files

An attempt to implement file storage.

Testing the minimum basic functionality.

Log in md for easy viewing.

LunexCoding committed:
Add major improvements

## Added

Added `IDGenerator.py` - performing recipe and ingredient.
 ID generation.
Added `filters.py` - filtering recipes by:
  * `name`
  * `rating`
Added tests. At the moment, work is underway to improve them, namely
 the creation as unit tests.

Added primary goals and objectives of this project.
 They are in the file `doc/GoalsAndObjectives.md`

## Changed

Changed `recipe.py`:
    *changed storage structure.

Changed `storage.py`:
  * Added a check for a duplicate product when adding a product
 to storage. If the given product already exists, the new quantity
 will be added to its quantity.
  * Products are now stored in `productID`: `amount` dictionary.

The `FileSystem.py` has been redesigned to provide full functionality
 of file operations. Improvements are possible.

Changed `ingredientsBD.py`:
  * Ingredients are now stored in the
 `ingredientID`: `amount` dictionary.
  * Added a check for a duplicate by the name of the ingredient in
 the database. If the name of the added ingredient is already in the
 database, the ingredient will not be added.

Changed `book.py`:
  * Recipes are now stored in `recipeID`:  `recipe obj` dictionary.
  * Added check for duplicate by recipe name in book. If the name
 of the recipe to be added is already specified in book, the recipe
 will not be added.
  * Added methods for reading and loading recipe files.

LunexCoding committed:
Add new unit tests

Added `book.py` tests.

Updated `ingredientDB.py` tests.

Filters are divided into Filters and Sortings.

Added Sorters `sorting.py`:
   * by name
   * by rating

Updated `FileSystem`.

oleksandr-grabovyi commented on IDGenerator.py:
```diff
@@ -0,0 +1,26 @@
+        self.__ingredientID = 0
+        self.__recipeID = 0
+
+    def loadLastIngredientID(self, lastID=0):
```
почему генератор ID что-то знает о ингредиентах? Это же максимально обобщенная сущность вообще. Если надо передавать какой-то диапазон - то так и назови методы. Ну типа restoreIDStates(lastID).

oleksandr-grabovyi commented on IDGenerator.py:
```diff
@@ -0,0 +1,26 @@
+    def loadLastRecipeID(self, lastID=0):
+        self.__recipeID = lastID
+
+    def getRecipeID(self):
```
а зачем разные группа ID еще и на рецепты?) Ну получается не IDGenerator, а какой-то SuperReceiptBookIDGeneratorFreeDownload. В нем должен в идеале быть 1 метод - generate() который вернет уникальную ID. Стартовое значение и т.д. - в конструктор. Если хочешь разные генераторы для разных иерархий - создаешь 2 генератора

oleksandr-grabovyi commented on RecipeBook/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .recipe import Recipe
```
тебе точно нужно это все? Эти импорты будут вызыватся при любом импорте модуля из пакета. Короче все что прописано тут будет при импорте модуля из этого пакета. Это стремно так то) Это не ошибка, просто если можно обойтись без - лучше обойтись без)

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,114 @@
+from sortings import BaseSorting, SortingByName, SortingByRating
+
+
+RECIPES_DIRECTORY = Path("data/recipes")
```
эти константы где-то еще используются? Если только в этом файле - поставь префикс  _

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,114 @@
+        self._recipes = {}
+        self._lastLoadedRecipeID = None
+        self._filters = {
+            'all': BaseFilter,
```
тут не классы сохраняй, а сразу сущности (созданные обьекты классов)

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,114 @@
+            'byName': SortingByName,
+            'byRating': SortingByRating,
+        }
+        self._activeFilter = self._filters['all']()
```
а тут просто вынимай готовую сущность

oleksandr-grabovyi replied:
тут вообще конечно хитрый момент. Если у тебя миллиард фильтров, а меняешь их крайне редко - то лучше так как написал ты, чтобы секонномить память на сущностях. Но если смена фильтра частая, то ты будешь часто создавать новую сущность вместо использования уже сохраненной. Тогда лучше то, что предложил я. Я думаю этот кейс более подходящий) хотя это не точно

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,114 @@
+            log.warning(f'Рецепт <{recipe.name}> уже есть в книге')
+
+    def _checkRecipeInBook(self, recipeName):
+        return True if recipeName not in [self.getRecipeByID(recipe).name for recipe in self._recipes] else False
```
слуш, а не лучше что-то типа
for recipe in self._recipes.values():     
    if recipe.name == recipeName:
        return True
return False

если через генератор хочешь, то хоть иди по values() словаря, а не ключам (for recipe in self._recipes.values():     )

oleksandr-grabovyi commented on RecipeBook/recipe.py:
```diff
@@ -0,0 +1,47 @@
+
+class Recipe:
+
+    def __init__(self, name=None, ingredients=None, cookingSteps=None, description=None, isFavorite=False, rating=0):
```
а они точно могут быть None? ) Как же имя, ингридиенты и т.д. будут None?

oleksandr-grabovyi commented on app.py:
```diff
@@ -0,0 +1,38 @@
+class App:
+
+    def __init__(self):
+       ...
```
это что такое? Аналог pass?

oleksandr-grabovyi commented on ingredientsDB.py:
```diff
@@ -0,0 +1,63 @@
+
+
+log = logger.getLogger(__name__)
+DATABASE_FILE = Path('data/ingredientsDB.json')
```
_DATABASE_FILE если только в этом файле используется

oleksandr-grabovyi commented on ingredientsDB.py:
```diff
@@ -0,0 +1,63 @@
+    def addIngredientToDB(self, ingredient):
+        if self._lastLoadedIngredientID is not None:
+            self._ingredients[int(self._lastLoadedIngredientID)] = ingredient
+        else:
```
elif?

oleksandr-grabovyi commented on jsonTools.py:
```diff
@@ -0,0 +1,17 @@
+
+
+class JsonEncoder(JSONEncoder):
+        def default(self, obj):
```
чет отступов многовато

oleksandr-grabovyi commented on logger.py:
```diff
@@ -0,0 +1,12 @@
+from Logger import Logger
+
+
+FILENAME = "log.md"
```
скрывай локальне константы с помощью профекса _. Ну ты понял)

oleksandr-grabovyi commented on tests/unit/test_book.py:
```diff
@@ -0,0 +1,90 @@
+        self._book.deleteRecipeByID(1)
+        log.debug(self._book.recipes)
+        self.assertNotIn(1, self._book.recipes)
+        self._book.addRecipe(self._recipe1)
```
а зачем эта строка уже после всех проверок?

oleksandr-grabovyi commented on tests/unit/test_book.py:
```diff
@@ -0,0 +1,90 @@
+        self.assertTrue(set(["2.json", "3.json"]).issubset(self._fileSystem.getFileNameWithSuffix(self._fileSystem.listDir(book.RECIPES_DIRECTORY))))
+
+    def test_loadRecipes(self):
+        self._book.loadRecipes()
```
эм. А что тут проверяешь то? Ну вызвал метод. А ожидаемое поведение какое-то?)) смысл теста?)

oleksandr-grabovyi commented on tests/unit/test_recipe.py:
```diff
@@ -0,0 +1,31 @@
+        )
+
+    def test_recipe(self):
+        self.assertEqual([self._recipe.name, self._recipe.rating, self._recipe.isFavorite], ['Капустница', 12, True])
```
совершенно не нужный тест) ничего он не дает толком)

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,114 @@
+            recipeIngredients = {}
+            for ingredientID, amount in recipe.ingredients.items():
+                recipeIngredients[int(ingredientID)] = amount
+            recipe.ingredients = recipeIngredients
```
а что тут вообще происходит? у тебя же такой же словарь как был, такой и будет

LunexCoding committed:
Update fileSystem

Module rewritten according to documentation https://docs.pyfilesystem.org/en/latest/

Added own exceptions for `fileSystem`

Added tests `fileSystem`

LunexCoding committed:
Change from numeric ID to UUID

Updated modules and tests that somehow affect the topic of ID

`storage` tests added

LunexCoding committed:
Fix some old bugs

oleksandr-grabovyi commented on RecipeBook/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from RecipeBook import book
+from .book import g_RecipeBook
```
что тебе дают тут импорты вообще? Это же не какая-то хитрая инциаилазция и т.д. Ну сможешь ты вместо
from RecipeBook.book import g_RecipeBook 
писать
from RecipeBook import g_RecipeBook
что оно даст? А вот вероятность провтыкать и получить циклический импорт растет сильно. Как и просто запутанность кода

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,125 @@
+    def __init__(self):
+        self._recipes = {}
+        self._filters = {
+            "all": baseFilter,
```
инстансы, не классы. baseFilter()

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,125 @@
+    def writeRecipesFiles(self):
+        for recipeID in self._recipes:
+            try:
+                self._writeRecipeFile(recipeID)
```
слуш, а у тебя рецепты по 1 пишутся? То есть ты каждый раз файл открываешь и закрываешь? Это не очень хорошо. Лучше чтобы запись в файл осуществлялась сразу списком рецептов. Ну если в списке всего 1 рецепт это ладно. Но если рецептов 100 надо сохранить, то открытие и закрытие файла это "дорогая" операция

oleksandr-grabovyi commented on RecipeBook/book.py:
```diff
@@ -0,0 +1,125 @@
+        _log.debug(f"Начата фильтрация {self._activeFilter.__class__}")
+        for recipe in self._recipes.values():
+            if self._activeFilter.call(recipe):
+                self.showRecipe(recipe)
```
аналогичино. Может сначала получить список рецептов, которые попали под фильтр, а потом все отобразить за 1 проход?

oleksandr-grabovyi commented on RecipeBook/recipe.py:
```diff
@@ -0,0 +1,46 @@
+        log.debug(f"Рецепт <{self._name}> создан")
+
+    def addIngredient(self, ingredientID, amount):
+        self._ingredients[ingredientID] = amount
```
наверное стоит сумировать. Если уже есть ингридиент, и я еще раз добавляю - то что считать правильным - последнее значение, или сумму? Я хз. Это вопрос дизайна программы, а не программирования если шо

oleksandr-grabovyi commented on app.py:
```diff
@@ -0,0 +1,36 @@
+            g_ingredientsDB.loadIngredients()
+            g_storage.loadProducts()
+            g_RecipeBook.loadRecipes()
+            g_RecipeBook.filtration()
```
а зачем тут фильтровать, если фильтрация в конце еще есть?

oleksandr-grabovyi commented on fileSystem.py:
```diff
@@ -0,0 +1,157 @@
+        if not path.is_dir():
+            raise IsNotDirectoryException(path)
+        if not len(os.listdir(path)) == 0:
+            raise IsNotEmptyException(path)
```
тут наверное стоит вернуть False, а не исключение) это же не ошибка. Пустая директория - возвращаем тру, не пустая - фолс. Если директории нет и т.д. - это исключение уже

oleksandr-grabovyi commented on fileSystem.py:
```diff
@@ -0,0 +1,157 @@
+        if (path.exists() and recreate is False):
+            raise PathExistsException(path)
+        path.mkdir(exist_ok=recreate)
+        return True
```
а тут зачем тру?

oleksandr-grabovyi commented on fileSystem.py:
```diff
@@ -0,0 +1,157 @@
+        if (path.exists() and path.is_dir()):
+            raise PathExistsAsDirectoryException(path)
+        path.unlink()
+        return True
```
а тут зачем тру?

oleksandr-grabovyi commented on fileSystem.py:
```diff
@@ -0,0 +1,157 @@
+        if not self.isEmpty(path):
+            raise IsNotEmptyException(path)
+        path.rmdir()
+        return True
```
а тут зачем тру?

oleksandr-grabovyi commented on fileSystem.py:
```diff
@@ -0,0 +1,157 @@
+            else:
+                self.removeTree(child)
+        path.rmdir()
+        return True
```
а тут зачем тру?

oleksandr-grabovyi commented on filters.py:
```diff
@@ -0,0 +1,28 @@
+        return True if all(listBool) else False
+
+
+baseFilter = _BaseFilter()
```
ой ой ой. Та наверное не надо создавать тут инстансы) Я понял почему там без скобок было ) Не-не. Сделай классы публичными, это убери. Не надо глобальную переменную сохранять. Там где фильтр понадобился - там и создавай сразу в словаре

oleksandr-grabovyi commented on ingredientsDB.py:
```diff
@@ -0,0 +1,66 @@
+        return self._ingredients
+
+
+g_ingredientsDB = IngredientsDatabase()
```
это аналог синглтона. Если так делаешь - пряч класс с помощью _

oleksandr-grabovyi commented on sortings.py:
```diff
@@ -0,0 +1,18 @@
+        return dict(sorted(recipes.items(), key=lambda x: x[1].rating, reverse=True))
+
+
+baseSorting = _BaseSorting()
```
аналогично как с фильтрами

oleksandr-grabovyi commented on storage.py:
```diff
@@ -0,0 +1,50 @@
+    def products(self):
+        return self._products
+
+g_storage = Storage()
```
_Storage()

oleksandr-grabovyi commented on storage.py:
```diff
@@ -0,0 +1,50 @@
+        return g_ingredientsDB.getIngredientByID(productID)
+
+    def getProductAmountByID(self, productID):
+        return self._products[productID]
```
self._products.get(productID, 0) может же ключ не валидным быть таки

oleksandr-grabovyi commented on storage.py:
```diff
@@ -0,0 +1,50 @@
+        return self._products[productID]
+
+    def deleteProductByID(self, productID):
+        del self._products[productID]
```
а если ключа нет, ошибку не выдаст?
